### PR TITLE
Feature: Sign-in as a browser user. Closes #2550

### LINKED ISSUE #2550: [Master Issue] Peer-to-peer referrals
Spec: https://docs.google.com/document/d/11_sUcZkdTsbsFuZNnyZila8vQM_IMqdEWR1srCVny8Y/edit

1. Signup via Uphold and promo creation flow on Publishers
2. Add support on eyeshade
3. Add support on Promo

### PR DESCRIPTION
**Pull request by yachtcaptain23**

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -575,6 +575,8 @@
     t.datetime "default_currency_confirmed_at"
     t.datetime "member_at"
     t.datetime "send_emails", default: -> { "CURRENT_TIMESTAMP" }
+    t.text "card_id"
```
Comment (corymcdonald) [inline db/schema.rb:578]: I don't think we need the card_id, this is what the `address` is.

File: test/controllers/api/v1/stats/promo_campaigns_controller_test.rb [hunk 2]
```diff
@@ -2,26 +2,20 @@
...
-    promo_registration2 = promo_registrations(:unattached_promo_registration)
     get "/api/v1/stats/promo_campaigns/", headers: { "HTTP_AUTHORIZATION" => "Token token=fake_api_auth_token" }
       assert_equal(
-        [{
-          promo_campaign_id: promo_campaign.id,
-          name: "test campaign",
-          promo_registrations: [
-            { promo_registration_id: promo_registration.id, referral_code: "PRO789" }
-          ]
+        PromoCampaign.all.map { |promo_campaign|
```
Comment (corymcdonald) [inline test/controllers/api/v1/stats/promo_campaigns_controller_test.rb:7]: We should change this to be a do/end block. I imagine the linter should have picked this up?

```suggestion
        PromoCampaign.all.map do |promo_campaign|
```

File: app/controllers/browser_users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class BrowserUsersController < ApplicationController
+  include PromosHelper
```
[the remark this answers is not in the document]
Comment (yachtcaptain23) [reply app/controllers/browser_users_controller.rb:2]: I believe it's for the promo partial in https://github.com/brave-intl/publishers/pull/2598/files#diff-88179ef1c52f2593281d9601e8403096R10

File: app/models/uphold_connection.rb [hunk 4]
```diff
@@ -9,6 +9,8 @@ class UpholdConnection < ActiveRecord::Base
   # Snooze for the next ~ 80 years, this is what I consider forever from now :)
   FOREVER_DATE = DateTime.new(2100, 1, 1)
 
+  USE_BROWSER = 1
```
[the remark this answers is not in the document]
Comment (yachtcaptain23) [reply app/models/uphold_connection.rb:12]: Nah, it's just a parameter for a switch case

File: db/schema.rb (hunk 1 above)
Comment (yachtcaptain23) [reply db/schema.rb:578]: `card_id` is the equivalent of `card_id` found on `uphold_connection_for_channels`.`card_id`.

File: app/controllers/browser_users_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (yachtcaptain23) [reply app/controllers/browser_users_controller.rb:2]: Nope, nvm I think you're right.

File: app/controllers/uphold_connections_controller.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    uphold_card_id = params[:uphold_card_id]
+    state = SecureRandom.hex(64).to_s
+    Rails.cache.write(state, uphold_card_id, expires_in: 10.minutes)
+    redirect_to Rails.application.secrets[:uphold_authorization_endpoint].
+      gsub('<UPHOLD_CLIENT_ID>', Rails.application.secrets[:uphold_login_client_id]).
+      gsub('<UPHOLD_SCOPE>', Rails.application.secrets[:uphold_scope]).
+      gsub('<STATE>', state)
+  end
+
+  def confirm
```
[the remark this answers is not in the document]
Comment (yachtcaptain23) [reply app/controllers/uphold_connections_controller.rb:13]: Looks great! I'll implement your change.

With regards to `return redirect_to`, I used the suggestion in the documentation https://www.rubydoc.info/docs/rails/3.1.1/AbstractController/DoubleRenderError which suggests `redirect_to and return`

File: app/controllers/uphold_connections_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    uphold_card_id = Rails.cache.fetch(params[:state])
+    uphold_connection, uphold_card = get_card(
+      uphold_card_id: uphold_card_id,
+      uphold_code: params[:code]
+    )
+
+    return redirect_to(root_path, flash: { alert: I18n.t(".out_of_time_alert") }) if uphold_card.blank?
```
Comment (corymcdonald) [inline app/controllers/uphold_connections_controller.rb:24]: Should this be `redirect_to ... and return` based on the comment you made above?

**Pull request merged**